File: src/main/java/io/github/faecraft/heartshapedbox/math/two_d/Square.java
package io.github.faecraft.heartshapedbox.math.two_d;

import net.minecraft.util.Pair;
import net.minecraft.util.math.Vec2f;

import java.util.Optional;

public class Square {
    final LineSegment a;
    final LineSegment b;
    final LineSegment c;
    final LineSegment d;
    
    final Vec2f min;
    final Vec2f max;
    
    public Square(Vec2f a, Vec2f b) {
        this.min = a;
        this.max = b;
        
        this.a = new LineSegment(a, new Vec2f(a.x, b.y));
        this.b = new LineSegment(new Vec2f(a.x, b.y), b);
        this.c = new LineSegment(new Vec2f(b.x, a.y), b);
        this.d = new LineSegment(a, new Vec2f(b.x, a.y));
    }
    
    public Pair<Vec2f[], Vec2f[]> splitFromLine(Line line) {
        Optional<Vec2f> aRes = a.intersectFromLine(line, false);
        if (aRes.isPresent()) {
            Optional<Vec2f> cRes = c.intersectFromLine(line, false);
            
            return new Pair<>(
                new Vec2f[]{
                    aRes.get(),
                    b.a,
                    max,
                    cRes.orElseThrow(IllegalStateException::new)
                },
                new Vec2f[]{
                    aRes.get(),
                    min,
                    d.b,
                    cRes.orElseThrow(IllegalStateException::new)
                }
            );
        } else {
            Optional<Vec2f> bRes = b.intersectFromLine(line, true);
            Optional<Vec2f> dRes = d.intersectFromLine(line, true);
    
            return new Pair<>(
                new Vec2f[]{
                    bRes.orElseThrow(IllegalStateException::new),
                    b.a,
                    min,
                    dRes.orElseThrow(IllegalStateException::new)
                },
                new Vec2f[]{
                    bRes.orElseThrow(IllegalStateException::new),
                    max,
                    d.b,
                    dRes.orElseThrow(IllegalStateException::new)
                }
            );
        }
    }
}

File: src/main/kotlin/io/github/faecraft/heartshapedbox/math/two_d/Square.kt
package io.github.faecraft.heartshapedbox.math.two_d

import net.minecraft.util.math.Vec2f
import java.util.*

class Square(a: Vec2f, b: Vec2f) {
    private val a: LineSegment = LineSegment(a, Vec2f(a.x, b.y))
    private val b: LineSegment = LineSegment(Vec2f(a.x, b.y), b)
    private val c: LineSegment = LineSegment(Vec2f(b.x, a.y), b)
    private val d: LineSegment = LineSegment(a, Vec2f(b.x, a.y))

    private val min: Vec2f = a
    private val max: Vec2f = b

    fun splitFromLine(line: Line): Pair<Array<Vec2f>, Array<Vec2f>> {
        val aRes: Optional<Vec2f> = a.intersectFromLine(line, false)
        return if (aRes.isPresent) {
            val cRes: Optional<Vec2f> = c.intersectFromLine(line, false)
            Pair(arrayOf(
                aRes.get(),
                b.a,
                max,
                cRes.orElseThrow { IllegalStateException() }
            ), arrayOf(
                aRes.get(),
                min,
                d.b,
                cRes.orElseThrow { IllegalStateException() }
            ))
        } else {
            val bRes: Optional<Vec2f> = b.intersectFromLine(line, true)
            val dRes: Optional<Vec2f> = d.intersectFromLine(line, true)
            Pair(arrayOf(
                bRes.orElseThrow { IllegalStateException() },
                b.a,
                min,
                dRes.orElseThrow { IllegalStateException() }
            ), arrayOf(
                bRes.orElseThrow { IllegalStateException() },
                max,
                d.b,
                dRes.orElseThrow { IllegalStateException() }
            ))
        }
    }
}
